### PR TITLE
kernel: Make ticks <-> ms/us conversion precise

### LINKED ISSUE #12509: Fix rounding in _ms_to_ticks()
At the moment the _ms_to_ticks() rounds down, which results in premature expiration of timeouts scheduled just before edge of tick. To fix that we have to revert https://github.com/zephyrproject-rtos/zephyr/pull/9745 in order to restore rounding up.

### PR DESCRIPTION
**Pull request by pizi-nordic**
The CONFIG_SYS_CLOCK_TICKS_PER_SEC does not represent
actual number of ticks per second due to rounding performed
in sys_clock_hw_cycles_per_tick(). As result the true length
of the tick could be accurately obtained only from
sys_clock_hw_cycles_per_tick() to sys_clock_hw_cycles_per_sec()
ratio.

This PR updates tick <-> ms/us conversion function eliminating
the CONFIG_SYS_CLOCK_TICKS_PER_SEC define from equation, thus
improving precision of the time unit conversion.

Fixes: #12509

### CODE REVIEW

File: include/sys_clock.h
Comment (jukkar) [inline include/sys_clock.h]: The `sys_clock_hw_cycles_per_tick()` looks suspicious, there are no ticks mentioned in other parts of this calculation.

Comment (pizi-nordic) [reply include/sys_clock.h]: The `sys_clock_hw_cycles_per_sec() / sys_clock_hw_cycles_per_tick` is the real number of ticks per second (if you take under account fraction part of the division). So to get number of ticks for given number of miliseconds, we have to multiply this value by `ms / MSEC_PER_SEC`.

**Review by carlescufi — DISMISSED** [review]
s/tics/ticks in commit message

**Review by carlescufi — COMMENTED** [review]
OK with me, but needs an approver that can examine the changes in-depth

**Comment by pizi-nordic** [comment]
@andyross: I don't expect that #17155 will go into 2.0, while this PR has a chance.
And this is the sole reason: As you said, the #17155 presents more comprehensive approach.

**Review by pabigot — COMMENTED** [review]
Ugh.

OK, so this reveals existing inconsistency in the current implementation: `__ticks_to_us` will do the "correct" thing given a negative tick count, while `__ticks_to_ms` will do the "wrong" thing given a negative tick count. The rework changes it so they behave consistently: both will get the "wrong" answer for negative values.  I haven't looked at what the function documentation prescribes.

Second, my recollection is that at least one test relaxed its condition in order to pass with an unfixed conversion bug, though I'm not sure it's this one.  If that is the case, any such tests should be revisited.

In any case, from a release management point of view I don't think this should go in as it may change the behavior of almost anything that depends on a tick-to-time translation.  I wouldn't block it 'cuz I'm not a release manager, but I won't approve it at this point.

I'd rather see this settled in #17155.

File: include/sys_clock.h
Comment (pizi-nordic) [reply include/sys_clock.h]: I don't think, that we are supposed to use this functions with negative timeouts. The sole place where `__ticks_to_us()` is used is here:
```
s32_t z_impl_k_usleep(int us)
{
        s32_t ticks;

        ticks = z_us_to_ticks(us);
        ticks = z_tick_sleep(ticks);
        return __ticks_to_us(ticks);
}
```

If we decide, that negative values should be supported here, we will have to change `__ticks_to_ms()` to behave correctly with the negative time.

**Comment by ioannisg** [comment]
Moving this to 2.1 since the review wa not done in time.

**Comment by pabigot** [comment]
@pizi-nordic Can we close this since #19591 supersedes #17155 and this (and solves #12509)?

**Comment by pizi-nordic** [comment]
IMHO  Yes.

**Pull request closed without merging**
